ops/model.py

```diff
@@ -43,7 +43,8 @@ def get_relation(self, relation_name, relation_id=None):
                     return relation
             else:
                 # The relation may be dead, but it is not forgotten.
-                return Relation(relation_name, relation_id, self._meta.relations[relation_name].role, self.unit, self._backend, self._cache)
+                our_app = self.app if self._meta.relations[relation_name].role == 'peers' else None
```

Comment:
The straight version of this is much easier to read:
```
if self._meta.relations[relation_name].role == 'peers'
    our_app = self.app
else:
    our_app = None
```
Also, this logic seems to be distributed in two different places without benefit. Instead of doing it here, and then having a new _meta attribute, and then handing the underlying relations value down into relations which is also stored, let's move all of this into the relations value itself. We might call an underlying private method in the RelationMapping value, but instead of:
```
model.get_relation(name, id)
```
I think we can simply have:
```
model.relations.get(name, id)
```
With id still defaulting to None. Seems nice and follows a recognized pattern.

Reply:
The problem with `model.relations.get(name, id)` is that the default mixin doesn't expect the second `id` argument so simply extending `__getitem__` with the second argument doesn't work because the relevant parameter is simply not passed down by the following code:

https://github.com/python/cpython/blob/3.7/Lib/_collections_abc.py#L657-L662

One way to work around this would be making the key compound.

```
self.relations[(name, id)]
```

But then we lose the simplicity of `self.relations[name]`. Supporting both brings up the thought of supporting `self.relations[id]` as well but we currently do not have any facilities to lookup just by a relation id without specifying a relation name as this requires doing a reverse lookup over all relation names to retrieve a relation name to be placed into the `RelationMap`.

Reply:
Here is a rework of the relation object lookup logic:
https://github.com/canonical/operator/pull/107/commits/127e1f19a2424c99dfafdc3ac43f33576df7a567

Nothing better came to mind today than supporting compound keys in `get`, however, let's discuss this tomorrow and see if there's a better way.

Reply:
Got rid of compound keys per our discussion with @niemeyer and @chipaca. See:
https://github.com/canonical/operator/pull/107/commits/38bd46da91afe6b13c733d880cfee3147d8c9393